instrgen/docs/design.md

```diff
@@ -0,0 +1,200 @@
+        // ID Dumps rewriter id.
+        Id() string
+        // Inject tells whether package should be rewritten.
+        Inject(pkg string, filepath string) bool
+        // ReplaceSource decides whether input sources should be replaced
+        // or all rewriting work should be done in temporary location.
+        ReplaceSource(pkg string, filePath string) bool
+        // Rewrite does actual package rewriting.
+        Rewrite(pkg string, file *ast.File, fset *token.FileSet, trace *os.File)
+        // WriteExtraFiles generate additional files that will be linked
+        // together to input package.
+        // Additional files have to be returned as array of file names.
+        WriteExtraFiles(pkg string, destPath string) []string
+}
+```
+
+
+#### Example rewriter
+
+```go
+// FunDumper rewrites all functions according to FilePattern.
+type FunDumper struct {
+}
+
+// Id.
+func (FunDumper) Id() string {
+        return "FunDumper"
+}
+
+// Inject.
+func (fd FunDumper) Inject(pkg string, filepath string) bool {
+	return true
+}
+
+// ReplaceSource.
+func (fd FunDumper) ReplaceSource(pkg string, filePath string) bool {
+        return true
+}
+
+func makeStmts(name string) []ast.Stmt
+	s := &ast.ExprStmt {
+		X: &ast.CallExpr {
+			Fun: &ast.SelectorExpr {
+				X: &ast.Ident {
+					Name: "__myfmt",
+				},
+				Sel: &ast.Ident {
+					Name: "Println",
+				},
+			},
+			Lparen: 42,
+			Args: []ast.Expr {
+				&ast.BasicLit {
+					ValuePos: 43,
+					Kind: STRING,
+					Value: "\" + name +\"",
+				},
+			},
+			Ellipsis: 0,
+		},
+	}
+        return []ast.Stmt{s}
+
+// Rewrite.
+func (fd FunDumper) Rewrite(pkg string, file *ast.File, fset *token.FileSet, trace *os.File) {
+        visited := make(map[string]bool, 0)
+        ast.Inspect(file, func(n ast.Node) bool {
+                if funDeclNode, ok := n.(*ast.FuncDecl); ok {
+                        // check if functions has been already instrumented
+                        if _, ok := visited[fset.Position(file.Pos()).String()+":"+funDeclNode.Name.Name]; !ok {
+                                funDeclNode.Body.List = append(makeStmts(funDeclNode.Name.Name), funDeclNode.Body.List...)
+                                astutil.AddNamedImport(fset, file, "__myfmt", "fmt")
+                                visited[fset.Position(file.Pos()).String()+":"+funDeclNode.Name.Name] = true
+                        }
+                }
+                return true
+        })
+}
+
+// WriteExtraFiles.
+func (FunDumper) WriteExtraFiles(pkg string, destPath string) []string {
+        return nil
+}
+```
+
+### UI
+
+To lower usage barrier to user `instrgen` offers simple UI where instrumentations of specific functions and libraries
+can be selected.
+
+![image info](./instrgen.png)
+
+### Scope
+
+First version will provide auto-instrumentation for following libraries:
+
+- HTTP
+- gRPC
+
+In the second phase we plan to add support for auto-instrumentation of selected user defined functions.
```

Comment:
I know this is the second phase, but i'm still not sure if this API will be less effort for users compared to just using the existing OTel APIs in their code.

Reply:
My plan to do that was via simple UI or configuration where you could select specific functions.